packages/backend-plugin-manager/src/scanner/plugin-scanner.ts

```diff
@@ -0,0 +1,302 @@
+          e,
+        );
+        continue;
+      }
+
+      let platform: PackagePlatform;
+      try {
+        platform = PackageRoles.getRoleInfo(
+          scannedPlugin.manifest.backstage.role,
+        ).platform;
+      } catch (e) {
+        this.logger.error(
+          `failed to load dynamic plugin manifest from '${pluginHome}'`,
+          e,
+        );
+        continue;
+      }
+
+      if (platform === 'node') {
+        if (this.preferAlpha) {
+          const pluginHomeAlpha = path.resolve(pluginHome, 'alpha');
+          if ((await fs.lstat(pluginHomeAlpha)).isDirectory()) {
+            const backstage = scannedPlugin.manifest.backstage;
+            try {
+              scannedPlugin = await this.scanDir(pluginHomeAlpha);
+            } catch (e) {
+              this.logger.error(
+                `failed to load dynamic plugin manifest from '${pluginHomeAlpha}'`,
+                e,
+              );
+              continue;
+            }
+            scannedPlugin.manifest.backstage = backstage;
+          } else {
+            this.logger.warn(
+              `skipping '${pluginHomeAlpha}' since it is not a directory`,
+            );
+          }
+        }
+      }
+
+      scannedPlugins.push(scannedPlugin);
+    }
+    return scannedPlugins;
+  }
+
+  private async scanDir(pluginHome: string): Promise<ScannedPluginPackage> {
+    const manifestFile = path.resolve(pluginHome, 'package.json');
+    const content = await fs.readFile(manifestFile);
+    const manifest: ScannedPluginManifest = JSON.parse(content.toString());
+    return {
+      location: url.pathToFileURL(pluginHome),
+      manifest: manifest,
+    };
+  }
+
+  async trackChanges(): Promise<void> {
+    const setupRootDirectoryWatcher = async (): Promise<void> => {
+      return new Promise((resolve, reject) => {
+        if (!this._rootDirectory) {
+          resolve();
+          return;
+        }
+        let ready = false;
+        this.rootDirectoryWatcher = chokidar
+          .watch(this._rootDirectory, {
+            ignoreInitial: true,
+            followSymlinks: true,
+          })
+          .on(
+            'all',
+            (
+              event: 'add' | 'addDir' | 'change' | 'unlink' | 'unlinkDir',
+              eventPath: string,
+              _: Stats | undefined,
+            ): void => {
+              if (
+                (['addDir', 'unlinkDir'].includes(event) &&
+                  path.dirname(eventPath) === this._rootDirectory) ||
+                (['add', 'unlink', 'change'].includes(event) &&
+                  path.dirname(path.dirname(eventPath)) ===
+                    this._rootDirectory &&
+                  path.basename(eventPath) === 'package.json')
+              ) {
+                this.logger.info(
+                  `rootDirectory changed (${event} - ${eventPath}): scanning plugins again`,
+                );
+                this.subscribers.forEach(s => s());
```

Comment:
Prolly want a bit of debounce here right? Or at least check for in-flight scans and just in general make sure that a lot of change events don't trigger a lot of reloads

Reply:
Fixed